src/source_sender/mod.rs

```diff
@@ -341,10 +403,17 @@ impl Inner {
     where
         E: Into<Event> + ByteSizeOf,
         I: IntoIterator<Item = E>,
+        <I as IntoIterator>::IntoIter: ExactSizeIterator,
     {
         let events = events.into_iter().map(Into::into);
+        let mut unsent_event_count = UnsentEventCount::new(events.len());
         for events in array::events_into_arrays(events, Some(CHUNK_SIZE)) {
-            self.send(events).await?;
+            let count = events.len();
+            self.send(events).await.map_err(|err| {
+                unsent_event_count.discard();
```

Comment:
I'm not clear why this `discard` is here, setting the count to zero. If the `self.send` returns with an error, aren't the events still unsent and should be reported as such?

Reply:
`discard` is here because the caller handles emitting a `StreamClosed` error, which increments `component_events_discarded_total`. I have a follow-up branch to consolidate that logic. For now, I'll add a comment.